src/succinctx/gateway.cairo

```diff
@@ -0,0 +1,343 @@
+                input.sha256(),
+                context.keccak(),
+                callback_addr,
+                callback_selector,
+                callback_gas_limit
+            );
+            self.requests.write(nonce, request_hash);
+            self
+                .emit(
+                    RequestCallback {
+                        nonce,
+                        function_id,
+                        input,
+                        context,
+                        callback_addr,
+                        callback_selector,
+                        callback_gas_limit,
+                        fee_amount: starknet::info::get_tx_info().unbox().max_fee,
+                    }
+                );
+
+            self.nonce.write(nonce + 1);
+
+            // TODO(#80): Implement Fee Vault and send fee
+
+            request_hash
+        }
+        /// Creates a proof request for a call. Equivalent to an off-chain request through an API.
+        ///
+        /// # Arguments
+        ///
+        /// * `function_id` - The function identifier.
+        /// * `input` - The function input.
+        /// * `entry_address` - The address of the callback contract.
+        /// * `entry_calldata` - The entry calldata for the call.
+        /// * `entry_gas_limit` - The gas limit for the call.
+        fn request_call(
+            ref self: ContractState,
+            function_id: u256,
+            input: Bytes,
+            entry_address: ContractAddress,
+            entry_calldata: Bytes,
+            entry_gas_limit: u32
+        ) {
+            self
+                .emit(
+                    RequestCall {
+                        function_id,
+                        input,
+                        entry_address,
+                        entry_calldata,
+                        entry_gas_limit,
+                        sender: starknet::info::get_caller_address(),
+                        fee_amount: starknet::info::get_tx_info().unbox().max_fee,
+                    }
+                );
+        // TODO(#80): Implement Fee Vault and send fee
+        }
+
+        /// If the call matches the currently verified function, returns the output. 
+        /// Else this function reverts.
+        ///
+        /// # Arguments
+        /// * `function_id` The function identifier.
+        /// * `input` The function input.
+        fn verified_call(self: @ContractState, function_id: u256, input: Bytes) -> (u256, u256) {
+            assert(self.verified_function_id.read() == function_id, Errors::INVALID_CALL);
+            assert(self.verified_input_hash.read() == input.sha256(), Errors::INVALID_CALL);
+
+            self.verified_output.read()
+        }
+
+        fn fulfill_callback(
+            ref self: ContractState,
+            nonce: u64,
+            function_id: u256,
+            input_hash: u256,
+            callback_addr: ContractAddress,
+            callback_selector: felt252,
+            callback_calldata: Span<felt252>,
```

Comment:
I see this is a new argument in our function ( compared to SuccinctGateway ). What were your thoughts on adding this argument instead of building the call data from `context` and `output` in the function:

Reply:
It appears solidity's `call` is a bit easier to work with in that it allows user to pack all calldata for call. 

Cairo's version of `call` requires the arguments to be explicitly declares `address`, `entry_point_selector`, and `calldata`.

This seemed to me to be an easier interface to accommodate the language difference.